zerver/tests/test_hooks.py

```diff
@@ -953,3 +953,42 @@ def test_yo_message(self):
 
         msg = Message.objects.filter().order_by('-id')[0]
         self.assertEqual(msg.content, (u"Yo from IAGO"))
+
+class TeamcityHookTests(AuthedTestCase):
+    def basic_test(self, type, expected_subject, expected_msg):
+        """
+        Build notifications are generated by Teamcity after build completes.
+        """
+
+        email = "hamlet@zulip.com"
+        api_key = self.get_api_key(email)
+        body = ujson.dumps(ujson.loads(self.fixture_data("teamcity", type)))
+
+        stream = "teamcity"
+        url = "/api/v1/external/teamcity?stream=%s&api_key=%s" % (stream, api_key)
+        self.subscribe_to_stream(email, stream)
+
+        self.client.post(url,
+                         body,
+                         stream_name=stream,
+                         content_type="application/json")
+
+        msg = Message.objects.filter().order_by('-id')[0]
+        self.assertEqual(msg.subject, expected_subject)
+        self.assertEqual(msg.content, expected_msg)
+
+    def test_teamcity_success(self):
+        self.basic_test("success", u"Project :: Compile", u"Project :: Compile build 5535 - CL 123456 was successful! :thumbsup:\nDetails: [changes](http://teamcity/viewLog.html?buildTypeId=Project_Compile&buildId=19952&tab=buildChangesDiv), [build log](http://teamcity/viewLog.html?buildTypeId=Project_Compile&buildId=19952)")
+
+    def test_teamcity_broken(self):
+        self.basic_test("broken", u"Project :: Compile", u"Project :: Compile build 5535 - CL 123456 is broken with status Exit code 1 (new)! :thumbsdown:\nDetails: [changes](http://teamcity/viewLog.html?buildTypeId=Project_Compile&buildId=19952&tab=buildChangesDiv), [build log](http://teamcity/viewLog.html?buildTypeId=Project_Compile&buildId=19952)")
+
+    def test_teamcity_failure(self):
+        self.basic_test("failure", u"Project :: Compile", u"Project :: Compile build 5535 - CL 123456 is still broken with status Exit code 1! :thumbsdown:\nDetails: [changes](http://teamcity/viewLog.html?buildTypeId=Project_Compile&buildId=19952&tab=buildChangesDiv), [build log](http://teamcity/viewLog.html?buildTypeId=Project_Compile&buildId=19952)")
+
+    def test_teamcity_fixed(self):
+        self.basic_test("fixed", u"Project :: Compile", u"Project :: Compile build 5535 - CL 123456 has been fixed! :thumbsup:\nDetails: [changes](http://teamcity/viewLog.html?buildTypeId=Project_Compile&buildId=19952&tab=buildChangesDiv), [build log](http://teamcity/viewLog.html?buildTypeId=Project_Compile&buildId=19952)")
+
+    def test_teamcity_personal(self):
+        self.basic_test("personal", "", u"Your personal build of Project :: Compile build 5535 - CL 123456 is broken with status Exit code 1 (new)! :thumbsdown:\nDetails: [changes](http://teamcity/viewLog.html?buildTypeId=Project_Compile&buildId=19952&tab=buildChangesDiv), [build log](http://teamcity/viewLog.html?buildTypeId=Project_Compile&buildId=19952)")
+
```

Comment:
Can you extend the test for the personal case to check that it sent a personal message?